### PR TITLE
Issue/8193 add new database connection pool setting

### PR DESCRIPTION
**Pull request by Hugo-Inmanta**
# Description

- [x] Add new settings for scheduler DB pool
- [x] Add new setting for the server DB pool
- [x] we default the new setting such that the server half of the old setting as default. Each scheduler gets 10% of the pool. 
- [ ] Document that every environment now needs a pool of its own
- [x] add also to upgrade notes 

closes https://github.com/inmanta/inmanta-core/issues/8193

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
~~- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)~~

### CODE REVIEW

**Review by sanderr — COMMENTED** [review]
I'm only partially in the loop for this one at the moment, I'm requesting Wouter instead.

@wouterdb if you don't have capacity for this, feel free to bounce it back.

File: src/inmanta/server/config.py [hunk 1]
```diff
@@ -44,13 +44,39 @@
...
+
+db_service_connection_pool_min_size = Option(
+    "database",
+    "service_connection_pool_min_size",
+    5,
+    "Number of connections the database connection pool will be initialized with",
+    is_int,
+)
+db_service_connection_pool_max_size = Option(
+    "database", "service_connection_pool_max_size", 35, "Max number of connections in the database connection pool", is_int
```
Comment (jptrindade) [inline src/inmanta/server/config.py:78]: Maybe we should mention that these are specific per environment (from what I understood, could be wrong)

**Review by jptrindade — APPROVED** [review]

File: src/inmanta/data/__init__.py [hunk 2]
```diff
@@ -2644,6 +2647,29 @@ def to_dto(self) -> m.Environment:
...
+            "the resource scheduler in this environment.",
+            validator=convert_int,
+        ),
+        SCHEDULER_DB_CONNECTION_TIMEOUT: Setting(
+            name=SCHEDULER_DB_CONNECTION_TIMEOUT,
+            typ="positive_float",
+            default=60.0,
+            doc="Database connection timeout in seconds.",
+            validator=convert_positive_float,
+        ),
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:2672]: I don't think we want to make this into environment settings. 

typically, the people with access to the DB setup have no access to the environment setting, and vice versa.

(also, setting these requires accessing the DB, so this is very risky business)

**Review by wouterdb — CHANGES_REQUESTED** [review]
see comment

File: changelogs/unreleased/8193-add-new-database-connection-pool-setting.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+change-type: major
+destination-branches: [master]
+sections:
+  upgrade-note: >
+    Previously, a single database connection pool was shared among the server and the agent, configured
+    via the ``database.connection_pool_max_size`` and ``database.connection_pool_min_size`` options.
+    For each environment, the new resource scheduler uses a database connection pool of its own.
+    Please review the following connection pool options and make sure they fit your
+    needs. Do keep in mind that these options will apply to all active environments:
+      - For the scheduler: ["agent.db_connection_timeout", agent.db_connection_pool_max_size", agent.db_connection_pool_min_size"]
```
Comment (wouterdb) [inline changelogs/unreleased/8193-add-new-database-connection-pool-setting.yml:12]: let's call it scheduler, not agent 

```suggestion
      - For the scheduler: ["scheduler.db_connection_timeout", scheduler.db_connection_pool_max_size", scheduler.db_connection_pool_min_size"]
```

File: changelogs/unreleased/8193-add-new-database-connection-pool-setting.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+      - For the server: [``database.service_connection_pool_max_size``, ``database.service_connection_pool_min_size``, ``database.connection_timeout``]
```
Comment (wouterdb) [inline changelogs/unreleased/8193-add-new-database-connection-pool-setting.yml:13]: I would format these as 

`server.xxx` where xx is the same as the one for the scheduler

File: src/inmanta/server/agentmanager.py [hunk 5]
```diff
@@ -1365,9 +1365,11 @@ async def _make_agent_config(
 name={opt.db_name.get()}
 username={opt.db_username.get()}
 password={opt.db_password.get()}
+service_connection_pool_min_size={agent_cfg.agent_db_connection_pool_min_size.get()}
+service_connection_pool_max_size={agent_cfg.agent_db_connection_pool_max_size.get()}
+service_connection_timeout={agent_cfg.agent_db_connection_timeout.get()}
```
Comment (wouterdb) [inline src/inmanta/server/agentmanager.py:1370]: This is unexpected?

I would expect the scheduler to use it's own settings, not those of the server?

File: src/inmanta/agent/config.py [hunk 6]
```diff
@@ -173,6 +173,31 @@
     is_lower_bounded_int(60),
 )
 
+agent_db_connection_pool_min_size: Option[int] = Option(
```
Comment (wouterdb) [inline src/inmanta/agent/config.py:176]: These need a default  that is derived from the actual value of the deprecated setting, not of the default of that setting.

File: src/inmanta/server/config.py [hunk 7]
```diff
@@ -44,13 +44,39 @@
 db_name = Option("database", "name", "inmanta", "The name of the database on the postgresql server", is_str)
 db_username = Option("database", "username", "postgres", "The username to access the database in the PostgreSQL server", is_str)
 db_password = Option("database", "password", None, "The password that belong to the database user", is_str)
+
 db_connection_pool_min_size = Option(
-    "database", "connection_pool_min_size", 10, "Number of connections the pool will be initialized with", is_int
+    "database",
+    "connection_pool_min_size",
+    10,
+    "[DEPRECATED USE :inmanta.config:option:`database.service_connection_pool_min_size`] Number of connections the database connection pool will be initialized with",
```
Comment (wouterdb) [inline src/inmanta/server/config.py:52]: There is a way to indicate deprecation that neatly links in the docs

Comment (Hugo-Inmanta) [reply src/inmanta/server/config.py]: Added `predecessor` to the new server options

File: src/inmanta/agent/config.py (hunk 6 above)
Comment (Hugo-Inmanta) [reply src/inmanta/agent/config.py]: Added dynamic defaults to pool size options for server and scheduler

File: src/inmanta/server/agentmanager.py (hunk 5 above)
Comment (Hugo-Inmanta) [reply src/inmanta/server/agentmanager.py]: Agreed, cleaned this up and refactored the `server_db_connect` function into `initialize_database_connection_pool` with proper arguments.

File: src/inmanta/server/config.py (hunk 1 above)
Comment (Hugo-Inmanta) [reply src/inmanta/server/config.py]: I cleaned this up a bit and mentioned for the scheduler settings that they are set for each environment.

File: src/inmanta/server/services/databaseservice.py [hunk 8]
```diff
@@ -133,23 +143,41 @@ async def _check_database_pool_exhaustion(self) -> None:
...
+        username=database_username,
+        password=database_password,
+        create_db_schema=create_db_schema,
         connection_pool_min_size=connection_pool_min_size,
         connection_pool_max_size=connection_pool_max_size,
         connection_timeout=connection_timeout,
     )
-    LOGGER.info("Connected to PostgreSQL database %s on %s:%d", opt.db_name.get(), database_host, database_port)
+    LOGGER.info("Connected to PostgreSQL database %s on %s:%d", database_name, database_host, database_port)
     return out
```
Comment (jptrindade) [inline src/inmanta/server/services/databaseservice.py:183]: What do you think about making these arguments optional, and if they are not provided, using the environment values?

**Review by wouterdb — APPROVED** [review]

File: src/inmanta/server/services/databaseservice.py (hunk 8 above)
Comment (Hugo-Inmanta) [reply src/inmanta/server/services/databaseservice.py:183]: Do you have a specific use case in mind ? I think these parameters are better off being set explicitly when starting up the server / the scheduler.

**Pull request closed without merging**